holoviews/core/data/pandas.py

```diff
@@ -75,16 +71,9 @@ def init(cls, eltype, data, kdims, vdims):
                 )
 
             # Handle reset of index if kdims reference index by name
-            for kd in kdims:
-                kd = dimension_name(kd)
-                if kd in data.columns:
-                    continue
-                if any(kd == ('index' if name is None else name)
-                       for name in index_names):
-                    data = data.reset_index()
-                    break
-
-            if kdims:
+            if len(kdims) == len(index_names) and {dimension_name(kd) for kd in kdims} == set(index_names):
+                pass
+            elif kdims:
```

Comment:
Maybe cleaner, maybe not

```suggestion
            if kdims and not (len(kdims) == len(index_names) and {dimension_name(kd) for kd in kdims} == set(index_names)):
```